// ipfs-cli/cjs/src/commands/dag/get.js
'use strict';

var parseDuration = require('parse-duration');
var toCidAndPath = require('ipfs-core-utils/to-cid-and-path');
var toString = require('uint8arrays/to-string');
var utils = require('../../utils.js');
var dagPB = require('@ipld/dag-pb');
var dagCBOR = require('@ipld/dag-cbor');
var raw = require('multiformats/codecs/raw');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      }
    });
  }
  n['default'] = e;
  return Object.freeze(n);
}

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);
var dagPB__namespace = /*#__PURE__*/_interopNamespace(dagPB);
var dagCBOR__namespace = /*#__PURE__*/_interopNamespace(dagCBOR);
var raw__namespace = /*#__PURE__*/_interopNamespace(raw);

var dagGet = {
  command: 'get <cid path>',
  describe: 'Get a dag node or value from ipfs.',
  builder: {
    'local-resolve': {
      type: 'boolean',
      default: false
    },
    'cid-base': {
      describe: 'Number base to display CIDs in.',
      type: 'string',
      default: 'base58btc'
    },
    'data-enc': {
      describe: 'String encoding to display data in.',
      type: 'string',
      choices: [
        'base16',
        'base64',
        'base58btc'
      ],
      default: 'base64'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default['default']
    }
  },
  async handler({
    ctx: {ipfs, print},
    cidpath,
    cidBase,
    dataEnc,
    localResolve,
    timeout
  }) {
    const options = {
      localResolve,
      timeout
    };
    const {cid, path} = toCidAndPath.toCidAndPath(cidpath);
    let result;
    try {
      result = await ipfs.dag.get(cid, {
        ...options,
        path
      });
    } catch (err) {
      return print(`dag get failed: ${ err }`);
    }
    if (options.localResolve) {
      print('resolving path within the node only');
      print(`remainder path: ${ result.remainderPath || 'n/a' }\n`);
    }
    const node = result.value;
    const base = await ipfs.bases.getBase(cidBase);
    if (cid.code === dagPB__namespace.code) {
      const dagNode = node;
      print(JSON.stringify({
        data: dagNode.Data ? toString.toString(node.Data, dataEnc) : undefined,
        links: (dagNode.Links || []).map(link => ({
          Name: utils.stripControlCharacters(link.Name),
          Size: link.Tsize,
          Cid: { '/': link.Hash.toString(base.encoder) }
        }))
      }));
    } else if (cid.code === raw__namespace.code) {
      print(toString.toString(node, dataEnc));
    } else if (cid.code === dagCBOR__namespace.code) {
      print(JSON.stringify(utils.makeEntriesPrintable(node, base)));
    } else {
      print(utils.escapeControlCharacters(node.toString()));
    }
  }
};

module.exports = dagGet;
